/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package problems;
import java.util.*;
import utils.*;
import java.io.IOException;

/**
 * @author tlvlp
 */
// class name should correspond to the rosalind problem code eg. FIB, GC, DNA..
public class mprt {

    /**
     * @param inList
     * @return
     */
    // main method for the solution of the rosalind problem in the class name
    
    public static String parser = "fasta";
    
    public static ArrayList<String> solve(ArrayList<String> inList) {
        ArrayList<String> outList = new ArrayList<>();
        //import the UniProt database
        ArrayList<String> UniProtDatabase = InputParser
        
        
        
        return outList;
    }
    
    public static boolean hasNGlycosylationMotif (String seq) {
    /*  To allow for the presence of its varying forms, a protein motif is represented by a shorthand as follows: 
        [XY] means "either X or Y" 
        {X} means "any amino acid except X." 
        For example, the N-glycosylation motif is written as 
        N{P}[ST]{P}.*/
        for (int i=0; i<seq.length()-3; i++) {
            
        }
    }
}